A51-custom-backend-metrics.md

```diff
@@ -0,0 +1,387 @@
+we can use callback-style APIs in a pull model to update the metrics data,
+however, this is suboptimal because it is not performance/scalability friendly.
+In a pull model, the time complexity of doing utilization measurement syscalls reaches O(M*N),
+M being the number of OOB clients and N being their frequency configuration.
+In a push model, the number of syscalls decouples with the number of OOB clients and their configurations.
+In fact, the server application is able to tune the performance by sampling the utilization data at a constant rate.
+
+It is by design that the OOB server sends the data regardless of whether it has been changed or 
+not since the last time it was sent. We considered a potential optimization, that is to skip 
+sending the unchanged metrics data since last response.
+To do that, we should not compare maps for de-duplication, instead, we can bump up a
+generation_id of the metrics data every time users make a change.
+Each client connection only needs to compare its previous generation_id to do deduplication.
+The optimization strategy can be implemented in the future if there are concrete use cases.
+
+In addition to the minimum report interval configuration at the OOB server, it might be useful
+in the future to also let the users specify the default report interval separately.
+The reporting frequency in the OOB request is final because it is a server-streaming API.
+Clients would terminate the OOB stream and reconnect if they want to change the frequency.
+Switching to a bidi stream will make it possible to change the reporting frequency in flight,
+but this can be done in the future if needed without impacting the public interface.
+
+## Implementation
+This will be implemented in Java, Go and C++.
+
+### Java
+#### Per-Query Reporting 
+
+At the server side, Java will provide a per-RPC metrics recorder for the user applications to record both utilization and
+request cost metrics. They can obtain a reference to the `CallMetricRecorder` using the static 
+`getCurrent()` method. All the methods are thread safe.
+
+```Java
+public final class CallMetricRecorder {
+  // Records a request cost metric measurement for the call.
+  public CallMetricRecorder recordRequestCostMetric(String name, double value);
+
+  // Records a utilization metric measurement for the call.
+  public CallMetricRecorder recordUtilizationMetric(String name, double value);
+
+  //Records the CPU utilization metric measurement for the call. 
+  public CallMetricRecorder recordCPUUtilizationMetric(double value);
+
+  // Records the memory utilization metric measurement for the call. 
+  public CallMetricRecorder recordMemoryUtilizationMetric(double value);
+
+  // Returns the call metric recorder attached to the current call.
+  public static CallMetricRecorder getCurrent();
+}
+```
+
+At the client side, Java will provide a utility function `OrcaClientStreamTracerFactory`. 
```

Comment:
I think we need to mention how it will avoid double-parsing. The Factory may be duplicated multiple times, but it modifies `StreamInfo` when creating a `ClientStreamTracer` instance to become aware of other instances of the Factory. The first will parse the header and the following ones will internally register a callback with the first.